### DISCUSSION TITLE
From 0 to glTF with WebGPU: Rendering the First glTF Mesh

### DISCUSSION THREAD

**Discussion by Twinklebear** [Twinklebear/webgpu-0-to-gltf#2] [General]
Ask any questions or post comments here for about the "Rendering the First glTF Mesh" post

**Comment by cmhhelgeson** [comment]
Hi, quick question, I was wondering what the primitive.positions.length property represents in the GLTFPrimitiveClass's render function. As I understand it, the positions property in the GLTFPrimitive Class represent a GLTFAccessor instance for the positions information in the buffer. However, in the current iteration of the GLTFAccessor Class, it doesn't seem like GLTFAccessor has a .length property, even though it is referenced when we are setting the VertexBuffer in GLTFPrimitive.render. Is this a mistake or am I missing something?

**Reply by cmhhelgeson** [reply]
I assume it should be this.positions.view.length

**Reply by cmhhelgeson** [reply]
I think the behavior of this.positions.length might just be undefined, I replaced that argument with undefined in my Typescript code, and although I don't have an object rendering yet, I'm no longer getting a buffer out of range error and the screen is correctly blitting the right color from my renderDescriptor colorAttachment.

**Reply by Twinklebear** [reply]
Good catch, you're right! I was debating if I should use TypeScript or JavaScript for this series (since the original code was just in JS), looks like I should have picked TypeScript 😄 .

I've fixed this in: https://github.com/Twinklebear/webgpu-0-to-gltf/commit/a9da73037bad4eada3a9848005cc1550e22dea4b , I use  `accessor.count * accessor.byteStride` since the view can be bigger than the region that the accessor references because the accessor can apply an additional byteOffset. I think using `view.length - accessor.byteOffset` should also be fine. I'll update the blog post as well.

**Reply by cmhhelgeson** [reply]
Thanks for the help 👍I'm actually working on the tutorial in Typescript, which is how I found the bug. Using this code for the typing of my JSON response: https://github.com/bwasty/gltf-loader-ts/blob/master/source/gltf.ts.

**Reply by cmhhelgeson** [reply]
Also there's a minor issue where you reference a MAT5 in your GLTFType object when it should be a MAT4

**Reply by cmhhelgeson** [reply]
I've followed the code pretty religiously but I still can't get anything to render.

**Reply by Twinklebear** [reply]
Whoops, I'll fix that typo

Have you put the code up on Github somewhere? I could take a quick skim through or try running it to see if something stands out. I might work up a typescript version of the code myself to check, after using TypeScript more recently I do prefer it over JavaScript.

**Reply by cmhhelgeson** [reply]
For sure, let me send it to you later tonight, I'm out for the day. I believe it's an issue with the shaders or the uniforms, usually when I get the colorAttachment without any additional errors or outputs, it means that something is wrong with that configuration.

**Reply by cmhhelgeson** [reply]
Apologies for sending this the morning after. I cloned WebGPU-Samples to use as a WebGPU playground, so my code is all contained within the gltf Sample and the glbUtils.ts file in utils. 
https://github.com/cmhhelgeson/webgpu-samples/blob/main/src/utils/glbUtils.ts
https://github.com/cmhhelgeson/webgpu-samples/tree/main/src/sample/gltfViewer

**Reply by Twinklebear** [reply]
No worries, I just dug into the code a bit now and it looks like there are two issues:

- The main issue looks to be here: https://github.com/cmhhelgeson/webgpu-samples/blob/main/src/utils/glbUtils.ts#L441 , it looks like for TypeScript (or just the DataView object) it should be `28 + jsonHeader.getUint32(12, true)`. That fixed the main issue I was seeing where the vertex/index/etc binary data was not right,  since the binary chunk was not created at the right offset.
- The default camera position is hiding the avocado, so in testing I also commented out applying the camera transform here: https://github.com/cmhhelgeson/webgpu-samples/blob/main/src/sample/gltfViewer/gltf.vert.wgsl#L19 , changing the code to:
   ```
  output.Position = vec4<f32>(input.position, 1.0);
   ```

with those changes I can get an image of the Avocado, though of course in a very small form since the camera is not positioned (fortunately, the avocado seems to render ok directly to normalized device coords).
<img width="451" alt="Screenshot 2023-07-26 at 5 51 14 PM" src="https://github.com/Twinklebear/webgpu-0-to-gltf/assets/1522476/08eed4a4-1092-4691-8d3c-7d2d35d115c3">

**Reply by cmhhelgeson** [reply]
Weird, why would the projectionMatrix be hiding the model?

**Reply by Twinklebear** [reply]
It could be that the projection+view matrix is not being computed or passed properly (so maybe it 0's out all vertices or makes them NaN or etc.), or simply that the camera is looking away from the model by default, or is zoomed out far enough that it's not visible.

**Reply by cmhhelgeson** [reply]
Came back to it and got it working.

https://github.com/Twinklebear/webgpu-0-to-gltf/assets/62450112/370ee5af-3cc5-4e14-8b6f-e698872736d8

**Reply by Twinklebear** [reply]
Looking good!
